src/kernels/gpu_reference_kernel/naive_conv.cpp

```diff
@@ -1233,6 +1507,137 @@ inline __device__ void naive_conv_wrw_ndhwc(const src_data_t* __restrict__ p_in,
+                                                      int n,
+                                                      int k_per_group,
+                                                      int c_per_group,
+                                                      int do_,
+                                                      int ho,
+                                                      int wo,
+                                                      int sz,
+                                                      int sy,
+                                                      int sx,
+                                                      int dz,
+                                                      int dy,
+                                                      int dx,
+                                                      int pz,
+                                                      int py,
+                                                      int px,
+                                                      int fz,
+                                                      int fy,
+                                                      int fx,
+                                                      int group)
+{
+    /*
+     *  need to compute total filter pixel: `group * k_per_group * fz * fy * fx
+     * * c_per_group`.
+     *  to distribute this workload, let one workgroup compute `fz * fy * fx *
+     * c_per_group` pixel,
+     *  hence need `group * k_per_group` workgroups (grid_size).
+     */
+    int k             = k_per_group * group;
+    int c             = c_per_group * group;
+    int thread_length = fz * fy * fx * c_per_group;
+    int bid           = blockIdx.x;
+    int ik            = bid % k_per_group;
+    int ig            = bid / k_per_group;
+
+    // p_in += static_cast<size_t>(ig) * c_per_group;
+    //
+    // p_wei += static_cast<size_t>(ig) * k_per_group * fz * fy * fx * c_per_group +
+    // static_cast<size_t>(ik) * fz * fy * fx * c_per_group;
+    //
+    // p_out += static_cast<size_t>(ig) * k_per_group + static_cast<size_t>(ik);
+
+    p_in += static_cast<size_t>(ig) * in_strides[1];
+
+    p_wei += static_cast<size_t>(ig) * wei_strides[5] + static_cast<size_t>(ik) * wei_strides[4];
+
+    p_out += static_cast<size_t>(ig) * out_strides[1] + static_cast<size_t>(ik) * out_strides[0];
+
+    for(int tid = threadIdx.x; tid < thread_length; tid += blockDim.x)
+    {
+        // TODO(Amber): these might need modification due to strides
+        int ic = tid % c_per_group;
+        int ix = (tid / c_per_group) % fx;
+        int iy = (tid / (c_per_group * fx)) % fy;
+        int iz = (tid / (c_per_group * fx * fy));
+
+        double value = .0f;
+
+        for(int in = 0; in < n; in++)
+        {
+            for(int ido = 0; ido < do_; ido++)
+            {
+                int valid_d = 1;
+                int cur_d   = sz * ido - pz + dz * iz;
+                if(cur_d < 0 || cur_d >= di)
+                    valid_d &= 0;
```

Comment:
I'd prefer it like this, but it's mostly a cosmetic change.
```suggestion
                int cur_d   = sz * ido - pz + dz * iz;
                bool valid_d = (cur_d >= 0 && cur_d < di);
```
The same for the others `valid_x`.

It still can be `int` through ternary operator, something like this: `int valid_d = (cur_d >= 0 && cur_d < di) ? 1 : 0;`

Reply:
you're right, these checks should be simplified to `bool` like you suggested but I'm going to hold off on it for this PR and implement it in a follow up PR so I can limit the amount of changes I do to the kernel (It will help in case there are bugs).

Reply:
will take up as a follow up PR.